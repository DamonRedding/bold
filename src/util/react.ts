import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
