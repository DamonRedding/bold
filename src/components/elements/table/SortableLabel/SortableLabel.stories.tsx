import { action } from '@storybook/addon-actions'
import { select, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { withPropTypes, withTheme } from '../../../../stories-addons'

import { SortableLabel, SortDirection } from './SortableLabel'

const dirOptions: { [key in SortDirection]: SortDirection } = {
    '': '',
    'ASC': 'ASC',
    'DESC': 'DESC',
}

storiesOf('Components/Table', module)
    .addDecorator(withPropTypes())
    .addDecorator(withTheme())
    .addDecorator(withKnobs)
    .add('SortableLabel', () => (
        <SortableLabel
            dir={select('dir', dirOptions, '')}
            onChange={action('onChange')}
        >
            Property
        </SortableLabel>
    ))
