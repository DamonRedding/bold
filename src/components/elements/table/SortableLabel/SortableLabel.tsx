import * as React from 'react'

import { withStyles, WithStylesProps } from '../../../../styles'
import { Icon } from '../../Icon/Icon'

export type SortDirection = 'ASC' | 'DESC' | ''

export interface SortableLabelProps extends WithStylesProps {
    direction: SortDirection
    onChange(sortDirection: SortDirection, shiftKey?: boolean): any
}

@withStyles
export class SortableLabel extends React.Component<SortableLabelProps> {
    render() {
        const { css, direction } = this.props
        const styles = {
            wrapper: {
                display: 'inline-flex',
                alignItems: 'center',
                cursor: 'pointer',
            },
        }

        const icon = (direction === 'ASC' && 'angleDown')
            || (direction === 'DESC' && 'angleUp')
            || 'sort'

        return (
            <span className={css(styles.wrapper)} onClick={this.handleClick}>
                {this.props.children}
                <Icon
                    style={{ marginLeft: '0.25rem' }}
                    color={icon === 'sort' ? 'disabled' : 'primary'}
                    size={1}
                    icon={icon}
                />
            </span>
        )
    }

    private handleClick = (event: React.MouseEvent<any>) => {
        this.props.onChange(toggleDirection(this.props.direction), event.shiftKey)
    }
}

export const toggleDirection = (dir: SortDirection): 'ASC' | 'DESC' => {
    return (dir === 'ASC' && 'DESC')
        || (dir === 'DESC' && 'ASC')
        || 'ASC'
}
