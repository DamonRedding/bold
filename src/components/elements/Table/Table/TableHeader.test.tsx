import { mount, render } from 'enzyme'
import * as React from 'react'

import { withTheme } from '../../../../test'
import { SortableLabel } from '../SortableLabel/SortableLabel'

import { Table, TableHead, TableHeader, TableRow } from './index'

describe('TableHeader', () => {
    it('should call onSortChange when clicked', () => {
        const sortChange = jest.fn()
        const wrapper = mount(withTheme(
            <Table>
                <TableHead>
                    <TableRow>
                        <TableHeader
                            sortable={true}
                            sortDirection='ASC'
                            onSortChange={sortChange}
                        >
                            Header #1
                        </TableHeader>
                    </TableRow>
                </TableHead>
            </Table>
        ))
        expect(sortChange).not.toHaveBeenCalled()
        wrapper.find(SortableLabel).simulate('click')
        expect(sortChange).toHaveBeenCalledWith('DESC', undefined)
    })
    it('should accept the style prop', () => {
        expect(render(withTheme(
            <Table>
                <TableHead>
                    <TableRow><TableHeader style={{ color: 'red' }}>Test</TableHeader></TableRow>
                </TableHead>
            </Table>
        ))).toMatchSnapshot()
    })
})
