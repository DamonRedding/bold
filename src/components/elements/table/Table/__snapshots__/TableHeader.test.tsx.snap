// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHeader should accept the style prop 1`] = `
.emotion-3 {
  width: 100%;
  border-collapse: collapse;
  font-size: 0.75rem;
  border: 1px solid #e6e6e6;
}

.emotion-0 {
  text-align: left;
  padding: 0.5rem 1rem;
  border-bottom: 1px solid #e6e6e6;
  color: red;
}

<table
  class="emotion-3"
>
  <thead
    class="emotion-1"
  >
    <tr
      class="emotion-1"
    >
      <th
        class="emotion-0"
      >
        Test
      </th>
    </tr>
  </thead>
</table>
`;
